nomad/task_group_host_volume_claim_endpoint.go

```diff
@@ -0,0 +1,179 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package nomad
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/go-memdb"
+	metrics "github.com/hashicorp/go-metrics/compat"
+	"github.com/hashicorp/nomad/acl"
+	"github.com/hashicorp/nomad/nomad/state"
+	"github.com/hashicorp/nomad/nomad/state/paginator"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+// TaskGroupHostVolumeClaim is the server RPC endpoint for task group volume claims
+type TaskGroupHostVolumeClaim struct {
+	srv    *Server
+	ctx    *RPCContext
+	logger hclog.Logger
+}
+
+func NewTaskGroupVolumeClaimEndpoint(srv *Server, ctx *RPCContext) *TaskGroupHostVolumeClaim {
+	return &TaskGroupHostVolumeClaim{srv: srv, ctx: ctx, logger: srv.logger.Named("task_group_host_volume_claim")}
+}
+
+func (tgvc *TaskGroupHostVolumeClaim) List(args *structs.TaskGroupVolumeClaimListRequest, reply *structs.TaskGroupVolumeClaimListResponse) error {
+	authErr := tgvc.srv.Authenticate(tgvc.ctx, args)
+	if done, err := tgvc.srv.forward(structs.TaskGroupHostVolumeClaimListRPCMethod, args, args, reply); done {
+		return err
+	}
+	tgvc.srv.MeasureRPCRate("task_group_volume_claim", structs.RateMetricList, args)
+	if authErr != nil {
+		return structs.ErrPermissionDenied
+	}
+	defer metrics.MeasureSince([]string{"nomad", "task_group_volume_claim", "list"}, time.Now())
+
+	allowClaim := acl.NamespaceValidator(acl.NamespaceCapabilityHostVolumeRead)
+	aclObj, err := tgvc.srv.ResolveACL(args)
+	if err != nil {
+		return err
+	}
+	if !allowClaim(aclObj, args.RequestNamespace()) {
+		return structs.ErrPermissionDenied
+	}
+
+	ns := args.RequestNamespace()
+
+	opts := blockingOptions{
+		queryOpts: &args.QueryOptions,
+		queryMeta: &reply.QueryMeta,
+		run: func(ws memdb.WatchSet, stateStore *state.StateStore) error {
+			var iter memdb.ResultIterator
+			var err error
+
+			switch {
+			case args.JobID != "":
+				iter, err = stateStore.TaskGroupHostVolumeClaimsByJobID(ws, args.JobID)
+			case args.TaskGroup != "":
+				iter, err = stateStore.TaskGroupHostVolumeClaimsByTaskGroup(ws, args.TaskGroup)
+			case args.VolumeName != "":
+				iter, err = stateStore.TaskGroupHostVolumeClaimsByVolumeName(ws, args.VolumeName)
```

Comment:
Minor optimization thoughts: 
* Should we move the volume name above the task group? It feels like you're more likely to check by volume name than by task group, and volume name is potentially a much smaller set than task group (which can repeat between jobs).
* Alternately, do we need the task group index? It feels unlikely that someone is going to search by task group along without job ID?